src/main/java/org/truffleruby/core/cast/BooleanCastNode.java

```diff
@@ -80,21 +75,32 @@ protected boolean doSymbol(RubySymbol object) {
     @Specialization(guards = "isForeignObject(object)", limit = "getCacheLimit()")
     protected boolean doForeignObject(Object object,
             @CachedLibrary("object") InteropLibrary objects,
-            @Cached ConditionProfile profile,
+            @Cached ConditionProfile isNullProfile,
+            @Cached ConditionProfile isBooleanProfile,
             @Cached BranchProfile failed) {
-        if (profile.profile(objects.isBoolean(object))) {
-            try {
-                return objects.asBoolean(object);
-            } catch (UnsupportedMessageException e) {
-                failed.enter();
-                // it concurrently stopped being boolean
+
+        if (isNullProfile.profile(objects.isNull(object))) {
+            return false;
+        } else {
+            if (isBooleanProfile.profile(objects.isBoolean(object))) {
+                try {
+                    return objects.asBoolean(object);
+                } catch (UnsupportedMessageException e) {
+                    failed.enter();
+                    // it concurrently stopped being boolean
+                    return true;
+                }
+            } else {
                 return true;
             }
-        } else {
-            return true;
         }
     }
 
+    @Specialization(guards = "isRubyDynamicObject(object)")
+    protected boolean doBasicObject(DynamicObject object) {
+        return true;
+    }
```

Comment:
We should keep this specialization higher than the foreign one, because it's expected to be more frequent than foreign objects. That benefits interpreter performance.

Reply:
Because we will find this specification earlier than foreign one?

Reply:
Yes